tutorials/Python/11_OOP/OOP.ipynb

```diff
@@ -0,0 +1,769 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<center>\n",
+    "<img src=\"SimpleMolecule.png\" alt=\"SimpleMolecule\" width=\"50%\"/>\n",
+    "</center>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "`SimpleMolecule` is now a class representing a molecule which can be initialised with elements and coordinates. In order to create an `acetylene` instance of the `SimpleMolecule` class we can do the following:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "acetylene = SimpleMolecule(\n",
+    "    els=np.array([1, 6, 6, 1]),\n",
+    "    coords=np.array([\n",
+    "        [-1.06, 0.00, 0.00], # H1\n",
+    "        [ 0.00, 0.00, 0.00], # C1\n",
+    "        [ 1.20, 0.00, 0.00], # C2\n",
+    "        [ 2.26, 0.00, 0.00], # H2\n",
+    "    ])\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "There are three important things to notice here. First, the `__init__` method takes `self` as the first of its three arguments. Second, the `__init__` function is not called explicitly but it's called implicitly as `SimpleMolecule(els, coords)`; this is why this method is \"special\" (some Pythonistas would call it _magic_!). Third, a `SimpleMolecule` object is created by passing two arguments and not three. The reason behind the latter behavior is that Python always passes, automatically, the object itself (which is an instance of the class `SimpleMolecule`) as first argument of a method. Therefore, when we initialise `SimpleMolecule` we only need to define the `els` and `coords` arguments, while the `self` argument is taken care of automatically by Python."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The `self` argument, which represents the particular instance of the class we are working with, can be used to store data related to that particular instance. The variables stored within a class are called _attributes_. In the case of `SimpleMolecule` the attributes are `self.elements` and `self.coordinates`. They can be easily accessed with the `.` operator:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [
+    {
+     "output_type": "stream",
+     "name": "stdout",
+     "text": [
+      "[1 6 6 1]\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(acetylene.elements)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "_Note #4_: The `acetylene` instance of the `SimpleMolecule` class is very similar to the dictionary we used before. With the dictionary we used `print(acetylene[\"elements\"])`, while now we use `print(acetylene.elements)`."
+   ]
+  },
+  {
+   "source": [
+    "> Create a new instance of `SimpleMolecule` representing a molecule of your choice. Print out elements and atomic coordinates."
```

Comment:
This is missing an empty box for students to put their code in.

Also we'll probably have to make a "solutions" copy with the answer.

Reply:
Is there an easy way to have hidden cells with the solutions instead of a copy (which is a maintainability nightmare)?

Reply:
I don't think so :( (or at least we had this discussion last year and I don't think we ever came up with an easy solution). I agree it's a real pain.

Reply:
@IAlibay How do you feel about using the [Exercise2 extension](https://jupyter-contrib-nbextensions.readthedocs.io/en/latest/nbextensions/exercise2/readme.html)?

Reply:
How likely is this not going to work on Windows or MacOS?

If we can guarantee that it'll install fine / play well, then sure.

Reply:
We also need to make sure this will run over binder in the one case where a student is unable to get conda working completly (I'm expecting there will be at least one person with an Apple M1 containing machine...)

Reply:
I think most extensions are in JS so they should work. I can give it a try on the OOP. If it works I could leave it there and then we can unify everything for next year.

Reply:
sounds like a good call 👍

Reply:
The extension `` works very well locally. I think if you don't have the extension installed, the cell with the solution just shows up as a normal cell (see [nbviewer](https://nbviewer.jupyter.org/github/bigginlab/OxCompBio/blob/458a26da031ddbcb2c88a497619e7c33c9299589/tutorials/Python/11_OOP/OOP.ipynb)). Since most of them is going to run things locally I guess it is ok?

We could leave it in this section only as a test and then migrate everything on a later date if we don't encounter any problem?

Reply:
Sounds good, do you want to raise an issue to remind us to look into implementing it in all the tutorials if we get a chance? (we can put it on the following priority list; setup > md stuff > merge solutions).